### PR TITLE
cffi: Support unnamed types and pass hi.go

### PR DESCRIPTION
**Pull request by corona10**
* Pass hi.go
* Support unnamed types.

This PR should be reviewd after #120 is merged.

### CODE REVIEW

File: bind/gencffi_func.go [hunk 1]
```diff
@@ -253,3 +254,76 @@ func (g *cffiGen) genFuncBody(f Func) {
...
+		fsym.goname,
+		strings.Join(funcArgs, ", "),
+		fsym.doc,
+	)
+	g.wrapper.Indent()
+	g._genFuncBody(sym, fsym)
+	g.wrapper.Outdent()
+}
+
+func (g *cffiGen) _genFuncBody(sym *symbol, fsym *symbol) {
```
Comment (sbinet) [inline bind/gencffi_func.go]: we don't like underscores :)

File: bind/gencffi_type.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+// license that can be found in the LICENSE file.
+
+package bind
+
+import (
+	"fmt"
+	"go/types"
+)
+
+// Generate Go type with Python wrapper class.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+	}
+
+	if sym.isSlice() {
+		g.genTypeIAdd(sym)
+	}
+
+	g.wrapper.Outdent()
+}
+
+// Generate Type __iadd__.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+func (g *cffiGen) genTypeIAdd(sym *symbol) {
+	g.wrapper.Printf("def __iadd__(self, value):\n")
+	g.wrapper.Indent()
+	switch {
+	case sym.isSlice():
+		g.wrapper.Printf("if not isinstance(value, collections.Sequence):\n")
```
Comment (sbinet) [inline bind/gencffi_type.go]: I don't think that's the test we want.

IIRC, python's `__iadd__` tests whether the value is iterable and complains if not:

```
$> python2
Python 2.7.13 (default, Jul 21 2017, 03:24:34) 
[GCC 7.1.1 20170630] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> lst=[]
>>> lst+=[1,2]
>>> lst
[1, 2]
>>> lst+=1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'int' object is not iterable
```

so I would test for `collections.Iterable` instead.
(and you actually do use the iteration protocol later on with `for elt in value:`)

File: bind/gencffi_type.go [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+		typ := sym.GoType().Underlying().(*types.Slice)
+		esym := g.pkg.syms.symtype(typ.Elem())
+		g.wrapper.Printf("for elt in value:\n")
+		g.wrapper.Indent()
+		g.wrapper.Printf("pyitem = _cffi_helper.cffi_%[1]s(elt)\n", esym.py2c)
+		g.wrapper.Printf("_cffi_helper.lib.cgo_func_%[1]s_append(self.cgopy, pyitem)\n", sym.id)
+		g.wrapper.Outdent()
+	default:
+		panic(fmt.Errorf(
+			"gopy: init for %s not handled",
```
Comment (sbinet) [inline bind/gencffi_type.go]: the error message looks like a copy-paste from the `init` handling :)

File: bind/gencffi_type.go [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+			sym.gofmt(),
+		))
+	}
+	g.wrapper.Printf("return self\n")
+	g.wrapper.Outdent()
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type __getitem__
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+		typ := sym.GoType().Underlying().(*types.Array)
+		esym := g.pkg.syms.symtype(typ.Elem())
+		g.wrapper.Printf("pyitem = _cffi_helper.cffi_%[1]s(item)\n", esym.c2py)
+	case sym.isSlice():
+		typ := sym.GoType().Underlying().(*types.Slice)
+		esym := g.pkg.syms.symtype(typ.Elem())
+		g.wrapper.Printf("pyitem = _cffi_helper.cffi_%[1]s(item)\n", esym.c2py)
+	default:
+		panic(fmt.Errorf(
+			"gopy: init for %s not handled",
```
Comment (sbinet) [inline bind/gencffi_type.go]: the error message looks like a copy-paste from the `__init__` handling.

File: bind/gencffi_type.go [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+			sym.gofmt(),
+		))
+	}
+	g.wrapper.Printf("return pyitem\n")
+	g.wrapper.Outdent()
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type __init__.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+		panic(fmt.Errorf(
+			"gopy: len for %s not handled",
+			sym.gofmt(),
+		))
+	}
+	g.wrapper.Outdent()
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type converter between C and Python.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+	g.wrapper.Printf("def cffi_cgopy_cnv_c2py_%[1]s(c):\n", sym.id)
+	g.wrapper.Indent()
+	g.wrapper.Printf("o = %[1]s()\n", typename)
+	g.wrapper.Printf("o.cgopy = c\n")
+	g.wrapper.Printf("return o\n")
+	g.wrapper.Outdent()
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type methods.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+					m.Name()+" || "+m.FullName(),
+				))
+			}
+			g._genFunc(sym, msym)
+		}
+	}
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type __repr__ method.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+	g.wrapper.Printf("def __repr__(self):\n")
+	g.wrapper.Indent()
+	g.wrapper.Printf("cret = _cffi_helper.lib.cgo_func_%[1]s_str(self.cgopy)\n", sym.id)
+	g.wrapper.Printf("ret = _cffi_helper.cffi_cgopy_cnv_c2py_string(cret)\n")
+	g.wrapper.Printf("return ret\n")
+	g.wrapper.Outdent()
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type __str__ method.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+	g.wrapper.Printf("def __str__(self):\n")
+	g.wrapper.Indent()
+	g.wrapper.Printf("cret = _cffi_helper.lib.cgo_func_%[1]s_str(self.cgopy)\n", sym.id)
+	g.wrapper.Printf("ret = _cffi_helper.cffi_cgopy_cnv_c2py_string(cret)\n")
+	g.wrapper.Printf("return ret\n")
+	g.wrapper.Outdent()
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type __setitem__
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+		g.wrapper.Printf("pyitem = _cffi_helper.cffi_%[1]s(item)\n", esym.py2c)
+		g.wrapper.Printf("_cffi_helper.lib.cgo_func_%[1]s_ass_item(self.cgopy, idx, pyitem)\n", sym.id)
+	case sym.isSlice():
+		typ := sym.GoType().Underlying().(*types.Slice)
+		esym := g.pkg.syms.symtype(typ.Elem())
+		g.wrapper.Printf("pyitem = _cffi_helper.cffi_%[1]s(item)\n", esym.py2c)
+		g.wrapper.Printf("_cffi_helper.lib.cgo_func_%[1]s_ass_item(self.cgopy, idx, pyitem)\n", sym.id)
+	default:
+		panic(fmt.Errorf(
+			"gopy: init for %s not handled",
```
Comment (sbinet) [inline bind/gencffi_type.go]: looks like a copy paste from `__init__`.

File: bind/symtab.go [hunk 15]
```diff
@@ -130,6 +130,10 @@ func (s symbol) isMap() bool {
 	return (s.kind & skMap) != 0
 }
 
+func (s symbol) isSequence() bool {
```
Comment (sbinet) [inline bind/symtab.go]: is this a function that tests for the Go-concept of a sequence or for the python concept of a sequence?

in other words, do we want to also test for maps ?

a quick grep in the code seems to indicate `isSequence` is used to determine whether we need to generate methods that implement the python's sequence protocol...

Comment (corona10) [inline bind/symtab.go]: > a quick grep in the code seems to indicate isSequence is used to determine whether we need to generate methods that implement the python's sequence protocol...
Yes, that is. :)
> in other words, do we want to also test for maps?
Yes, we need to generate protocols for map also. I am investigating about that.

**Comment by corona10** [comment]
@sbinet Updated!

File: bind/symtab.go (hunk 15 above)
Comment (corona10) [inline bind/symtab.go]: @sbinet  `isIterable` might be better?

Comment (sbinet) [inline bind/symtab.go]: I think `isPySequence` better conveys what piece of information we want to extract and its later use.

**Comment by corona10** [comment]
@sbinet Done PTAL!

File: bind/gencffi_type.go [hunk 16]
```diff
@@ -68,7 +85,36 @@ func (g *cffiGen) genTypeIAdd(sym *symbol) {
...
 }
 
-// genTypeInit generates Type __init__.
+// genTypeGetItem generates __getitem__ of a type.
+func (g *cffiGen) genTypeGetItem(sym *symbol) {
+	g.wrapper.Printf("def __getitem__(self, idx):\n")
+	g.wrapper.Indent()
+	g.wrapper.Printf("if idx >= len(self):\n")
+	g.wrapper.Indent()
+	g.wrapper.Printf("raise IndexError('array index out of range')\n")
```
Comment (sbinet) [inline bind/gencffi_type.go]: would it be possible to raise `"array index ..."` when it's a Go array and `"slice index ..."` when it's a Go slice ?

File: bind/gencffi_type.go [hunk 17]
```diff
@@ -196,3 +281,32 @@ func (g *cffiGen) genTypeTPStr(sym *symbol) {
...
 	g.wrapper.Printf("\n")
 }
+
+// genTypeSetItem generates __setitem__ of a type.
+func (g *cffiGen) genTypeSetItem(sym *symbol) {
+	g.wrapper.Printf("def __setitem__(self, idx, item):\n")
+	g.wrapper.Indent()
+	g.wrapper.Printf("if idx >= len(self):\n")
+	g.wrapper.Indent()
+	g.wrapper.Printf("raise IndexError('array index out of range')\n")
```
Comment (sbinet) [inline bind/gencffi_type.go]: ditto: it would be much nicer to raise an error with the correct type name (array/slice).

File: main_test.go [hunk 18]
```diff
@@ -342,6 +342,136 @@ len(slice): 2
...
+--- p.Work(2)...
+working...
+worked for 2 hours
+--- p.Work(24)...
+working...
+caught: can't work for 24 hours!
+--- p.Salary(2): 20
+--- p.Salary(24): caught: can't work for 24 hours!
+--- Person.__init__
+caught: invalid type for '%!s(BADINDEX)' attribute | err-type: <type 'exceptions.TypeError'>
```
Comment (sbinet) [inline main_test.go]: that seems like an error somewhere in our generated code...

File: bind/gencffi_struct.go [hunk 19]
```diff
@@ -103,6 +104,10 @@ func (g *cffiGen) genStructInit(s Struct) {
 		g.wrapper.Outdent()
 		g.wrapper.Printf("if %[1]s != None:\n", kwd_name)
 		g.wrapper.Indent()
+		g.wrapper.Printf("if not isinstance(%[1]s, %[2]s):\n", kwd_name, ifield.sym.pysig)
+		g.wrapper.Indent()
+		g.wrapper.Printf("raise TypeError(\"invalid type for '%[2]s' attribute\")\n", field.Name())
```
Comment (sbinet) [inline bind/gencffi_struct.go]: the index should be `%[1]s` instead.

File: bind/gencffi_type.go (hunk 16 above)
Comment (corona10) [inline bind/gencffi_type.go]: @sbinet That looks great I will update it.

**Comment by corona10** [comment]
@sbinet Updated and sorry for some mistakes.

File: main_test.go [hunk 20]
```diff
@@ -342,6 +342,136 @@ len(slice): 2
...
+mem(arr): caught: cannot make memory view because object does not have the buffer interface
+--- testing slice...
+slice: []int{1, 2}
+len(slice): 2
+slice[0]: 1
+slice[1]: 2
+slice[2]: caught: slice index out of range
+slice: []int{1, 42}
+len(slice): 2
+mem(slice): caught: cannot make memory view because object does not have the buffer interface
```
Comment (sbinet) [inline main_test.go]: ok.
please file an issue so we don't forget to implement the buffer interface for slices.

File: main_test.go [hunk 21]
```diff
@@ -342,6 +342,136 @@ len(slice): 2
...
+--- testing GC... [ok]
+--- testing array...
+arr: [2]int{1, 2}
+len(arr): 2
+arr[0]: 1
+arr[1]: 2
+arr[2]: caught: array index out of range
+arr: [2]int{1, 42}
+len(arr): 2
+mem(arr): caught: cannot make memory view because object does not have the buffer interface
```
Comment (sbinet) [inline main_test.go]: ok.
please file an issue so we don't forget to implement the buffer interface for arrays.

**Pull request merged**